pallets/gas/src/property_tests.rs

```diff
@@ -0,0 +1,212 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2022 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Properties and invariants that are checked:
+//! 1. all non-external nodes have a parent in GasTree storage.
+//! 2. all non-external nodes have at least one ancestor with value (i.e., ValueNode::node_with_value procedure always return Ok),
+//! however this value can be equal to 0. Also there are no guarantees that this ancestor is a parent.
+//! 3. all nodes can't have consumed parent with zero refs (there can't be any nodes like that in storage) between calls to `ValueTree::consume`.
+//! Therefore, if node is deleted, it is consumed and has zero refs (and zero value, if it is able to hold value and is of `ValueType::SpecifiedLocal` type).
+//! So if there is an existing consumed node, it has non-zero refs counter and a value >= 0 (between calls to `ValueTree::consume`)
+//! 4. if a sub-tree of `GasTree` is a tree, where root has `ValueType::External` type, then sub-tree's root is always deleted last.
+//! 5. nodes can become consumed only after `GasTree::consume` call.
+//! 6. Unspec refs counter for the current node is incremented only after `GasTree::split` which creates a node with `ValueType::UnspecifiedLocal` type.
+//! 7. Spec refs counter for the current node is incremented only after `GasTree::split_with_value`, which creates a node with `ValueType::SpecifiedLocal` type is.
+
+use super::*;
+use crate::mock::*;
+use frame_support::assert_ok;
+use primitive_types::H256;
+use proptest::prelude::*;
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    iter::FromIterator,
+};
+
+use strategies::GasTreeAction;
+use utils::RingGet;
+
+mod assertions;
+mod strategies;
+mod utils;
+
+type Gas = Pallet<Test>;
+
+proptest! {
+    #![proptest_config(ProptestConfig::with_cases(100))] // todo change to 1500
```

Comment:
I think it would be nice, to open an issue for setting test-cases/running proptests in accordance to run/test profile.

Currently will set test cases to 600, because:
1. The more test cases - the more probability to find an error.
2. The more test cases - the better shrinking works in case some assertion doesn't work.

Reply:
https://github.com/gear-tech/gear/issues/1080